File: src/main/java/arc/arc/stock/StockPlayer.java
package arc.arc.stock;

import arc.arc.configs.StockConfig;
import arc.arc.hooks.HookRegistry;
import arc.arc.network.repos.RepoData;
import arc.arc.util.TextUtil;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.*;
import net.kyori.adventure.text.minimessage.tag.Tag;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;

import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;

import static arc.arc.util.TextUtil.mm;

@AllArgsConstructor
@NoArgsConstructor
@ToString
@Setter
@Getter
public class StockPlayer extends RepoData<StockPlayer> {

    String playerName;
    UUID playerUuid;
    Map<String, List<Position>> positionMap = new HashMap<>();
    private double balance = 0;
    boolean autoTake = true;
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    double totalGains = 0;
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    double receivedDividend = 0;

    public StockPlayer(String name, UUID uuid) {
        this.playerUuid = uuid;
        this.playerName = name;
        setDirty(true);
    }

    @JsonIgnore
    public synchronized void addToBalance(double add, boolean fromPosition) {
        balance += add;
        if (fromPosition) totalGains += add;
        setDirty(true);
    }

    @JsonIgnore
    public synchronized double totalBalance() {
        double fromPositions = positionMap.values().stream()
                .flatMap(list -> list.stream().map(Position::totalValue))
                .reduce(0.0, Double::sum);
        //System.out.println("From positions: "+fromPositions);
        return balance + fromPositions;
    }

    @JsonIgnore
    public synchronized double giveDividend(String symbol) {
        if (!positionMap.containsKey(symbol)) return 0;
        System.out.println("Giving divedend to "+playerName);
        Stock stock = StockMarket.stock(symbol);
        if (stock.dividend < 0.00001) return 0;

        setDirty(true);

        double gave = 0;
        for (Position position : positionMap.get(symbol)) {
            double dividend = stock.dividend * position.amount;
            if (dividend == 0) continue;
            balance += dividend;
            receivedDividend += dividend;
            gave+=dividend;
            position.setReceivedDividend(position.getReceivedDividend()+dividend);
        }
        return gave;
    }

    @JsonIgnore
    public synchronized Optional<Position> find(String symbol, UUID uuid) {
        if (!positionMap.containsKey(symbol)) return Optional.empty();
        return positionMap.get(symbol).stream().filter(p -> p.positionUuid.equals(uuid)).findAny();
    }

    @JsonIgnore
    public boolean isBelowMaxStockAmount(){
        int currentAmount = positions().size();
        if(currentAmount < StockConfig.defaultStockMaxAmount) return true;
        var entry = StockConfig.permissionMap.ceilingEntry(currentAmount+1);
        if(entry == null) return false;
        OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(playerUuid);
        if(!offlinePlayer.isOnline() && HookRegistry.luckPermsHook == null) return false;
        return HookRegistry.luckPermsHook.hasPermission(offlinePlayer, entry.getValue());
    }

    @JsonIgnore
    public int maxStockAmount(){
        int max = -1;
        OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(playerUuid);
        if(!offlinePlayer.isOnline() && HookRegistry.luckPermsHook == null) return -1;
        for(var entry : StockConfig.permissionMap.entrySet()){
            if(HookRegistry.luckPermsHook.hasPermission(offlinePlayer, entry.getValue())
                    && entry.getKey() > max) max = entry.getKey();
        }
        return max == -1 ? StockConfig.defaultStockMaxAmount : max;
    }

    @JsonIgnore
    public TagResolver tagResolver() {
        double bal = this.getBalance();
        double total = this.totalBalance();
        return TagResolver.builder()
                .resolver(TagResolver.resolver("balance", Tag.inserting(
                        mm(TextUtil.formatAmount(bal), true)
                )))
                .resolver(TagResolver.resolver("name", Tag.inserting(
                        mm(playerName, true)
                )))
                .resolver(TagResolver.resolver("position_amount", Tag.inserting(
                        mm(positions().size()+"", true)
                )))
                .resolver(TagResolver.resolver("uuid", Tag.inserting(
                        mm(playerUuid.toString().split("-")[0], true)
                )))
                .resolver(TagResolver.resolver("auto_take", Tag.inserting(
                        mm(autoTake ? "<green>Да" : "<red>Нет", true)
                )))
                .resolver(TagResolver.resolver("received_dividends", Tag.inserting(
                        mm(TextUtil.formatAmount(receivedDividend), true)
                )))
                .resolver(TagResolver.resolver("total_balance", Tag.inserting(
                        mm(TextUtil.formatAmount(total), true)
                )))
                .resolver(TagResolver.resolver("gains", Tag.inserting(
                        mm(TextUtil.formatAmount(total-bal), true)
                )))
                .resolver(TagResolver.resolver("total_gains", Tag.inserting(
                        mm(TextUtil.formatAmount(getTotalGains()), true)
                )))
                .resolver(TagResolver.resolver("position_count", Tag.inserting(
                        mm(positions().size() + "", true)
                )))
                .build();
    }

    @JsonIgnore
    public synchronized Optional<Position> remove(String symbol, UUID uuid) {
        if (!positionMap.containsKey(symbol)) {
            //System.out.println("No key for symbol "+symbol);
            return Optional.empty();
        }
        for (Position position : positionMap.get(symbol)) {
            if (position.positionUuid.equals(uuid)) {
                setDirty(true);
                positionMap.get(symbol).remove(position);
                return Optional.of(position);
            }
        }
        return Optional.empty();
    }

    @JsonIgnore
    public synchronized void addPosition(Position position) {
        if (position == null) {
            System.out.println("Position is null!");
            return;
        }
        setDirty(true);
        positionMap.putIfAbsent(position.symbol, new CopyOnWriteArrayList<>());
        positionMap.get(position.symbol).add(position);
    }

    public void setAutoTake(boolean autoTake) {
        if (autoTake == this.autoTake) return;
        setDirty(true);
        this.autoTake = autoTake;
    }


    @JsonIgnore
    public synchronized List<Position> positions(String symbol) {
        return positionMap.get(symbol);
    }

    @JsonIgnore
    public synchronized List<Position> positions() {
        return positionMap.values().stream().flatMap(Collection::stream).toList();
    }

    @JsonIgnore
    public synchronized double totalGains() {
        return positionMap.values().stream()
                .flatMap(list -> list.stream().map(Position::gains))
                .reduce(0.0, Double::sum);
    }
    @JsonIgnore
    public Player player() {
        OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(playerUuid);
        if(offlinePlayer instanceof Player p) return p;
        return null;
    }

    @Override
    public String id() {
        return playerUuid.toString();
    }

    @Override
    public boolean isRemove() {
        return false;
    }

    @Override
    public synchronized void merge(StockPlayer other) {
        if (other == null) return;
        if (other.balance != 0) balance = other.balance;
        if (other.autoTake != autoTake) autoTake = other.autoTake;
        if (other.totalGains != 0) totalGains = other.totalGains;
        if (other.receivedDividend != 0) receivedDividend = other.receivedDividend;
        if (other.positionMap != null){
            positionMap.clear();
            positionMap.putAll(other.positionMap);
        }
    }
}

File: src/main/kotlin/ru/arc/stock/StockPlayer.kt
package ru.arc.stock

import net.kyori.adventure.text.minimessage.tag.Tag
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import ru.arc.config.StockConfig
import ru.arc.hooks.HookRegistry
import ru.arc.repository.Entity
import ru.arc.util.Logging.debug
import ru.arc.util.TextUtil
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

class StockPlayer(
    var playerName: String = "",
    var playerUuid: UUID = UUID.randomUUID(),
    var positionMap: MutableMap<String, MutableList<Position>> = HashMap(),
    private var balance: Double = 0.0,
    var autoTake: Boolean = true,
    var totalGains: Double = 0.0,
    var receivedDividend: Double = 0.0,
) : Entity {

    fun getBalance(): Double = balance

    fun setBalance(value: Double) {
        balance = value
    }

    @Synchronized
    fun addToBalance(add: Double, fromPosition: Boolean) {
        balance += add
        if (fromPosition) totalGains += add
    }

    @Synchronized
    fun totalBalance(): Double {
        val fromPositions = positionMap.values
            .flatMap { list -> list.map { it.totalValue() } }
            .fold(0.0, Double::plus)
        return balance + fromPositions
    }

    @Synchronized
    fun giveDividend(symbol: String): Double {
        if (!positionMap.containsKey(symbol)) return 0.0
        debug("Giving dividend for {} to {}", symbol, playerName)
        val stock = StockMarket.stock(symbol) ?: return 0.0
        if (stock.dividend < 0.00001) return 0.0

        var gave = 0.0
        for (position in positionMap[symbol]!!) {
            val dividend = stock.dividend * position.amount
            if (dividend == 0.0) continue
            balance += dividend
            receivedDividend += dividend
            gave += dividend
            position.receivedDividend += dividend
        }
        return gave
    }

    @Synchronized
    fun find(symbol: String, uuid: UUID): Position? =
        positionMap[symbol]?.firstOrNull { it.positionUuid == uuid }

    fun isBelowMaxStockAmount(): Boolean {
        val currentAmount = positions().size
        if (currentAmount < StockConfig.defaultStockMaxAmount) return true
        val entry = StockConfig.permissionMap.ceilingEntry(currentAmount + 1) ?: return false
        val offlinePlayer = Bukkit.getOfflinePlayer(playerUuid)
        if (!offlinePlayer.isOnline && HookRegistry.luckPermsHook == null) return false
        return HookRegistry.luckPermsHook!!.hasPermission(offlinePlayer, entry.value)
    }

    fun maxStockAmount(): Int {
        var max = -1
        val offlinePlayer = Bukkit.getOfflinePlayer(playerUuid)
        if (!offlinePlayer.isOnline && HookRegistry.luckPermsHook == null) return -1
        for (entry in StockConfig.permissionMap.entries) {
            if (HookRegistry.luckPermsHook!!.hasPermission(offlinePlayer, entry.value) && entry.key > max) {
                max = entry.key
            }
        }
        return if (max == -1) StockConfig.defaultStockMaxAmount else max
    }

    fun tagResolver(): TagResolver {
        val bal = balance
        val total = totalBalance()
        return TagResolver.builder()
            .resolver(TagResolver.resolver("balance", Tag.inserting(TextUtil.mm(TextUtil.formatAmount(bal), true))))
            .resolver(TagResolver.resolver("name", Tag.inserting(TextUtil.mm(playerName, true))))
            .resolver(TagResolver.resolver("position_amount", Tag.inserting(TextUtil.mm("${positions().size}", true))))
            .resolver(TagResolver.resolver("uuid", Tag.inserting(TextUtil.mm(playerUuid.toString().split("-")[0], true))))
            .resolver(TagResolver.resolver("auto_take", Tag.inserting(TextUtil.mm(if (autoTake) "<green>Да" else "<red>Нет", true))))
            .resolver(TagResolver.resolver("received_dividends", Tag.inserting(TextUtil.mm(TextUtil.formatAmount(receivedDividend), true))))
            .resolver(TagResolver.resolver("total_balance", Tag.inserting(TextUtil.mm(TextUtil.formatAmount(total), true))))
            .resolver(TagResolver.resolver("gains", Tag.inserting(TextUtil.mm(TextUtil.formatAmount(total - bal), true))))
            .resolver(TagResolver.resolver("total_gains", Tag.inserting(TextUtil.mm(TextUtil.formatAmount(totalGains), true))))
            .resolver(TagResolver.resolver("position_count", Tag.inserting(TextUtil.mm("${positions().size}", true))))
            .build()
    }

    @Synchronized
    fun remove(symbol: String, uuid: UUID): Position? {
        val list = positionMap[symbol] ?: return null
        val position = list.firstOrNull { it.positionUuid == uuid } ?: return null
        list.remove(position)
        if (list.isEmpty()) positionMap.remove(symbol)
        return position
    }

    @Synchronized
    fun addPosition(position: Position) {
        positionMap.getOrPut(position.symbol) { CopyOnWriteArrayList() }.add(position)
    }

    fun updateAutoTake(newValue: Boolean) {
        if (newValue == autoTake) return
        autoTake = newValue
    }

    @Synchronized
    fun positions(symbol: String): List<Position>? = positionMap[symbol]

    @Synchronized
    fun positions(): List<Position> = positionMap.values.flatten()

    @Synchronized
    fun totalGainsList(): Double = positionMap.values
        .flatMap { it.map(Position::gains) }
        .fold(0.0, Double::plus)

    fun player(): Player? {
        val offlinePlayer = Bukkit.getOfflinePlayer(playerUuid)
        return offlinePlayer as? Player
    }

    override fun id(): String = playerUuid.toString()
}
